src/cdi_CPEloss/Tabular_CP_Eloss.cc

```diff
@@ -0,0 +1,270 @@
+  }
+
+  // Get first temperature support point
+  nlines = static_cast<int>(
+      std::ceil(static_cast<double>(n_temperature) / max_entries));
+  line_entries = read_line();
+  min_log_temperature = stod(line_entries[0]);
+  for (uint32_t n = 0; n < nlines - 1; n++) {
+    read_line();
+  }
+  temperatures.resize(n_temperature);
+  for (uint32_t n = 0; n < n_temperature; n++) {
+    temperatures[n] = exp(min_log_temperature + n * d_log_temperature);
+  }
+
+  std::vector<double> stopping_data_1d(n_energy * n_density * n_temperature);
+
+  bool target_found = false;
+  nlines = static_cast<int>(std::ceil(
+      ((double)n_energy * n_density * n_temperature) /
+      max_entries)); // The number of lines taken up by stopping power data for one target
+  if (target.get_zaid() == -1) {
+    // Target is free electrons
+    target_found = true;
+    uint32_t nentry = 0;
+    for (uint32_t n = 0; n < nlines; n++) {
+      line_entries = read_line();
+      for (std::string entry : line_entries) {
+        stopping_data_1d[nentry] = stod(entry);
+        nentry++;
+      }
+    }
+  } else {
+    // Skip electrons
+    for (uint32_t n = 0; n < nlines; n++) {
+      line_entries = read_line();
+    }
+
+    // Find ion target, if it exists
+    const uint32_t n_target_ions =
+        stoi(read_line()[0]); // Number of target ions in file
+    for (uint32_t n_target_ion = 0; n_target_ion < n_target_ions;
+         n_target_ion++) {
+      int zaid_target_ion = stoi(read_line()[0]); // ZAID
+      read_line();                                // Z, A, mass
+      if (zaid_target_ion == target.get_zaid()) {
+        // This is the requested target ion
+        target_found = true;
+        uint32_t nentry = 0;
+        for (uint32_t n = 0; n < nlines; n++) {
+          line_entries = read_line();
+          for (std::string entry : line_entries) {
+            stopping_data_1d[nentry] = stod(entry);
+            nentry++;
+          }
+        }
+        break;
+      } else {
+        // This is not the requested target ion
+        for (uint32_t n = 0; n < nlines; n++) {
+          read_line();
+        }
+      }
+    }
+  }
+  file.close();
+
+  std::ostringstream dedx_error;
+  dedx_error << "Error finding target ZAID \"" << target.get_zaid()
+             << "\" in DEDX file \"" << filename << "\"";
+  Insist(target_found, dedx_error.str());
+
+  stopping_data.resize(n_energy, n_density, n_temperature);
+  for (uint32_t ne = 0; ne < n_energy; ne++) {
+    for (uint32_t nd = 0; nd < n_density; nd++) {
+      for (uint32_t nt = 0; nt < n_temperature; nt++) {
+        stopping_data(ne, nd, nt) =
+            stopping_data_1d[ne + n_energy * (nd + n_density * nt)];
+      }
+    }
+  }
+
+  // Convert units on table to match those of getEloss:
+  //   energy:      MeV -> cm/shk (using target particle mass)
+  double energy_cgs = exp(min_log_energy) * (1.e6 * 1.6021772e-12);
+  min_log_energy = log(sqrt(2. * energy_cgs / target.get_mass()) * 1.e-8);
```

Comment:
Let's talk about this unit conversion.  We might want to use entries from the `units` package to ensure we use the same conversion factors in all of our code.  The `units` package has some known deficiencies that @along and I are discussing, but I think it is better to use it for now.

Reply:
Looking at `units`, I'm not sure that module supports what we want to do here: converting from a particle energy to a particle speed. I might be able to convert particle energy in eV to Kelvin and then to ergs with Boltzmann, but that feels unsatisfactory.